eth_utils/abi.py

```diff
@@ -1,68 +1,793 @@
-    # Whatever comes after "tuple" is the array dims. The ABI spec states that
-    # this will have the form "", "[]", or "[k]".
-    array_dim = typ[5:]
-    collapsed = f"({delimited}){array_dim}"
+    # convert NamedTuple to regular tuple
+    typing = tuple if isinstance(aligned_arg, tuple) else type(aligned_arg)
 
-    return collapsed
+    return typing(
+        _align_abi_input(sub_abi, sub_arg)
+        for sub_abi, sub_arg in zip(sub_abis, aligned_arg)
+    )
+
+
+def _get_tuple_type_str_and_dims(s: str) -> Optional[Tuple[str, Optional[str]]]:
+    """
+    Takes a JSON ABI type string.  For tuple type strings, returns the separated
+    prefix and array dimension parts.  For all other strings, returns ``None``.
+    """
+    tuple_type_str_re = "^(tuple)((\\[([1-9]\\d*\b)?])*)??$"
+    match = re.compile(tuple_type_str_re).match(s)
+
+    if match is not None:
+        tuple_prefix = match.group(1)
+        tuple_dims = match.group(2)
+
+        return tuple_prefix, tuple_dims
+
+    return None
+
+
+def abi_to_signature(abi_element: ABIElement) -> str:
+    """
+    Returns a string signature representation of the function or event ABI
+    and arguments.
+
+    Signatures consist of the name followed by a list of arguments.
+
+    :param abi_element: ABI element.
+    :type abi_element: `ABIElement`
+    :return: Stringified ABI signature
+    :rtype: `str`
+
+    .. doctest::
+
+        >>> from eth_utils import abi_to_signature
+        >>> abi_element = {
+        ...   'constant': False,
+        ...   'inputs': [
+        ...     {
+        ...       'name': 's',
+        ...       'type': 'uint256'
+        ...     }
+        ...   ],
+        ...   'name': 'f',
+        ...   'outputs': [],
+        ...   'payable': False,
+        ...   'stateMutability': 'nonpayable',
+        ...   'type': 'function'
+        ... }
+        >>> abi_to_signature(abi_element)
+        'f(uint256)'
+    """
+    if abi_element["type"] == "fallback" or abi_element["type"] == "receive":
+        name = str(abi_element["type"])
+        abi_element_inputs = None
+    elif abi_element["type"] == "constructor":
+        name = abi_element["type"]
+        abi_element_inputs = abi_element["inputs"]
+    elif abi_element["type"] == "event":
+        name = abi_element["name"]
+        abi_element_inputs = abi_element["inputs"]
+    else:
+        name = str(abi_element.get("name", abi_element["type"]))
+        abi_element_inputs = abi_element.get("inputs")
+
+    return f"{name}({_abi_input_types(abi_element_inputs)})"
+
+
+def filter_abi_by_name(
+    name: str, contract_abi: ABI
+) -> List[Union[ABIFunction, ABIEvent, ABIError]]:
+    """
+    Get one or more function and event ABIs by name.
+
+    :param name: Name of the function or event.
+    :type name: `str`
+    :param contract_abi: Contract ABI.
+    :type contract_abi: `ABI`
+    :return: Function or event ABIs with matching name.
+    :rtype: `List[ABIElement]`
```

Comment:
This is different than the actual return type (`ABIElement` includes `ABICallable` instead of `ABIFunction`)